File: Machine Learning A-Z New/Part 2 - Regression/Section 4 - Simple Linear Regression/simple_linear_regression.py
# Simple Linear Regression - Salary_Data.csv
# get current working directory Spyder
import os
os.getcwd()

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# =============================================================================
# import data and setup Independent X and dependent y variables
# =============================================================================

# Importing the dataset
dataset = pd.read_csv('Salary_Data.csv')
# x will be the array that makes up the independent variable(s) years of experience
# reading the [] bit we are saying to the left of the , that we want all the variables
# on the right of the , we are saying all of them except the one on the right (-1)
X = dataset.iloc[:, :-1].values
# y will be the array that makes up the dependent variable in this case Salary
# 0,1,2,3 so 1 is the Salary column and the dependent variable
y = dataset.iloc[:, 1].values

# scatterplot of data
plt.scatter(X,y)

# =============================================================================
# missing data - - NO MISSING DATA THIS TIME
# =============================================================================

# =============================================================================
#  split data Train and Test sets and scatterplot
# =============================================================================

# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 1/3, random_state = 0)
# scatterplot of Train and Test data
plt.plot(X_train,y_train, 'go', label='Train Data')
plt.plot(X_test,y_test, 'b*', label='Test Data')
plt.title('Train and Test data Scatterplot')
plt.xlabel('X axis Independent Variable')
plt.ylabel('Y axis Dependent Variable')
plt.legend(loc='best')  # legend text comes from the plot's label parameter.
plt.show()

# =============================================================================
# Feature Scaling - not necessary for this model, the variables are OK without
# =============================================================================
# also the linear model function will take care of this for us

"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
sc_y = StandardScaler()
y_train = sc_y.fit_transform(y_train.reshape(-1,1))"""

# =============================================================================
# Fitting our Model to the Training set 
# =============================================================================

# Fitting Simple Linear Regression to the Training set
from sklearn.linear_model import LinearRegression
?LinearRegression
# now we create an object of the LinearRegression test
# regressor is our 'Machine' in the idea of Machine Learning
regressor = LinearRegression()
# then apply the function 'fit' on X and y_train, interestingly all at the same time
# this is slightly different than fit_transform
# here we 'train' our 'Machine' on our data, the 'Machine' 'Learned'
regressor.fit(X_train, y_train)

# =============================================================================
# Predicting the Test set results
# =============================================================================
# vector of predictions for the dependent variables
# again we use the regressor object that we trained above
# place cursor at end of predict and command i to see help
# in this case we used out Machine regressor to predict salaries from the Texst data
y_pred = regressor.predict(X_test)

# now we'll see how well we did
# Visualising the Train set results against the model's prediction
plt.scatter(X_train, y_train, color = 'red', label='Train Data')
# regression line of the predictions of the Train set of data 
plt.plot(X_train, regressor.predict(X_train), color = 'blue', label='Regression of X_train')
plt.title('Salary vs Experience (Training set)')
plt.xlabel('Years of Experience')
plt.ylabel('Salary')
plt.legend(loc='best')
plt.show()

# Visualising the Test set results against the models predictions (against Train)
plt.scatter(X_test, y_test, color = 'red', label='Test Data')
# regression line
plt.plot(X_train, regressor.predict(X_train), color = 'blue', label='Regression of X_train')
plt.title('Salary vs Experience (Test set)')
plt.xlabel('Years of Experience')
plt.ylabel('Salary')
plt.show()

# Visualising the Training set results
plt.scatter(X_train, y_train, color = 'red', label='Train Data')
plt.scatter(X_test,y_test, color = 'blue', label='Test Data' )
# regression line 
plt.plot(X_train, regressor.predict(X_train), color = 'green', label='Regression of X_train')
plt.title('Salary vs Experience (Training & Test set)')
plt.xlabel('Years of Experience')
plt.ylabel('Salary')
plt.legend(loc='best')
plt.show()

File: Machine Learning A-Z New/Part 2 - Regression/Section 4 - Simple Linear Regression/simple_linear_regression.R
# Simple Linear Regression
getwd()
setwd('/Users/markloessi/Machine_Learning/Machine Learning A-Z New/Part 2 - Regression/Section 4 - Simple Linear Regression')

# ======= Importing the dataset
dataset = read.csv('Salary_Data.csv')

# ======= MISSING DATA - none
# ======= FACTORING - we have no categorical data

# ======= Splitting the dataset into the Training set and Test set
# install.packages('caTools')
library(caTools)
set.seed(123)
# dependent variable is Salary
split = sample.split(dataset$Salary, SplitRatio = 2/3)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)
# we will train our model on the Training data and the model will learn the correlations
# between number of years experience (Independent Variable) and Slarary (Dependent Variable)
# then we'll test the Model on the Test data

# ======= Feature Scaling - as in Python no need to feature scale here 
# the process we'll use and the data here do not require feature scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

# ======= Fitting Simple Linear Regression to the Training set
# we will setup the model 'regressor'
# the 'formula' say that Salary (the dependent variable) is proportional to the YearsExperience
# 'data' is the training_set which we want to build the model on
regressor = lm(formula = Salary ~ YearsExperience,
               data = training_set)
# see some details of the regressor
summary(regressor)
# the good stuff, we can see that the P value (with the *** is highly significant)
# Coefficients:
#                    Estimate Std. Error t value Pr(>|t|)    
#   (Intercept)      27658.6     2632.0   10.51 4.14e-09 ***
#   YearsExperience   9275.6      403.9   22.97 8.74e-15 ***

# ======= Predicting the Test set results
# creating a vector of predicted values for the test set using the fitted model
y_pred = predict(regressor, newdata = test_set)
y_pred

# ======= Visualising the Training set results
# install.packages('ggplot2')
library(ggplot2)
# install.packages('tidyverse')
library(tidyverse)
ggplot() +
  geom_point(aes(x = training_set$YearsExperience, y = training_set$Salary),
             colour = 'red') +
  # regression line
  geom_line(aes(x = training_set$YearsExperience, y = predict(regressor, newdata = training_set)),
            colour = 'blue') +
  ggtitle('Salary vs Experience (Training set Data)') +
  xlab('Years of experience') +
  ylab('Salary') 
?ggplot
# ======= Visualising the Test set results
ggplot() +
  geom_point(aes(x = test_set$YearsExperience, y = test_set$Salary),
             colour = 'red') +
  geom_line(aes(x = training_set$YearsExperience, y = predict(regressor, newdata = training_set)),
            colour = 'blue') +
  ggtitle('Salary vs Experience (Test set)') +
  xlab('Years of experience') +
  ylab('Salary')




